File: Data Structures & Algorithms/top-k-elements-in-list/submission-1.java
//Bucket sort
class Solution {
    public int[] topKFrequent(int[] nums, int k) {
        Map<Integer,Integer> freqMap = new HashMap<>();
        //This is to handle the case where the nums array contains all of the same elements: Eg: [8,8,8,8,8] in this case we need an array with index=5
        List<Integer>[] freq = new List[nums.length+1];
        for (int num: nums){
            freqMap.merge(num,1,Integer::sum);
        }
        for (int i=0;i<freq.length;i++){
            freq[i] = new ArrayList<>();
        }
        //Mapping the frequency count to the number (remember why we do this)
        for(Map.Entry<Integer,Integer> entry: freqMap.entrySet()){
            freq[entry.getValue()].add(entry.getKey());
        }
        int[] res = new int[k];
        int index=0;
        //This is just for production safety
        for(int i=freq.length-1;i>0;i--){
            for(int n: freq[i]){
                res[index++] = n;
                if(index==k){
                    return res;
                }
            }
        }
    return res;

    }
}

File: Data Structures & Algorithms/products-of-array-discluding-self/submission-1.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        n = len(nums)

        res = [1] * n

        for i in range(1,n):
            res[i] = nums[i-1] * res[i-1]

        ans = nums[n-1]
        for i in range(n-2,-1,-1):
            res[i] = res[i] * ans
            ans = ans * nums[i]

        return res